examples/albef/model.py

```diff
@@ -0,0 +1,385 @@
+            input_ids=start_ids,
+            attention_mask=atts,
+            encoder_hidden_states=encoder_outputs.multimodal_embeddings,
+            encoder_attention_mask=question_atts,
+        )
+
+        logits = prediction_scores[:, 0, :]  # get the cls token
+
+        answer_first_token = answers[:, 1]
+        prob_first_token = F.softmax(logits, dim=1).index_select(
+            dim=1, index=answer_first_token
+        )
+        topk_probs, topk_ids = prob_first_token.topk(k, dim=1)
+
+        input_ids = []
+        input_atts = []
+
+        for topk_id in topk_ids:
+            input_ids.append(answers.index_select(dim=0, index=topk_id))
+            input_atts.append(answer_atts.index_select(dim=0, index=topk_id))
+        input_ids = torch.cat(input_ids)
+        input_atts = torch.cat(input_atts)
+        targets_ids = input_ids.masked_fill(input_ids == 0, -100)
+
+        question_states = encoder_outputs.multimodal_embeddings.repeat_interleave(
+            k, dim=0
+        )
+        question_atts = question_atts.repeat_interleave(k, dim=0)
+
+        prediction_scores = self.answer_decoder(
+            input_ids=input_ids,
+            attention_mask=input_atts,
+            encoder_hidden_states=question_states,
+            encoder_attention_mask=question_atts,
+        )
+
+        answer_loss = self.loss(targets_ids, prediction_scores)
+        answer_loss = answer_loss.view(input_ids.size(0), -1)
+
+        # topk_prob: first token probability
+        topk_probs = topk_probs.view(-1, 1)
+        log_probs = torch.cat([topk_probs.log(), -answer_loss], dim=1)
+
+        # re-calculate log probabilities for the answer sequences using chain rule
+        log_probs_sum = log_probs.sum(1)
+        log_probs_sum = log_probs_sum.view(num_ques, k)
+
+        topk_probs = F.softmax(log_probs_sum, dim=-1)
+
+        # get top-k after re-ranking
+        topk_probs, rerank_id = topk_probs.topk(k, dim=1)
+        topk_ids = torch.gather(topk_ids, 1, rerank_id)
+
+        return topk_ids, topk_probs
+
+    def _encoder_hidden_states(
+        self, multimodal_embeds, multimodal_embeds_m, question_atts, ans_lengths
+    ):
+        encoder_hidden_states = []
+        encoder_attention_mask = []
+        for b, n in enumerate(ans_lengths):
+            encoder_hidden_states += [multimodal_embeds[b]] * n
+            encoder_attention_mask += [question_atts[b]] * n
+        encoder_hidden_states = torch.stack(encoder_hidden_states)
+        encoder_attention_mask = torch.stack(encoder_attention_mask)
+
+        with torch.no_grad():
+            encoder_hidden_states_m = []
+            for b, n in enumerate(ans_lengths):
+                encoder_hidden_states_m += [multimodal_embeds_m[b]] * n
+            encoder_hidden_states_m = torch.stack(encoder_hidden_states_m)
+
+        return encoder_hidden_states, encoder_hidden_states_m, encoder_attention_mask
+
+
+def albef_vision_encoder(
+    hidden_size: int,
+    args: dict,
```

Comment:
Just curious, why do we treat hidden_size differently than all the other args in these builders?

Reply:
Other args are encoder-specific, but `hidden_size` should be the same for the 3 encoders so the hidden states can be passed around.